GUACAMOLE-626: Add support for Docker secrets to startup.sh

WojoInc opened the pull request:
Did my best to follow the CONTRIBUTING, as I meant to submit these changes a while back.
JIRA: https://issues.apache.org/jira/browse/GUACAMOLE-626

mike-jumper commented on guacamole-docker/bin/start.sh:
```diff
@@ -141,16 +141,39 @@ environment variables:
     MYSQL_DATABASE     The name of the MySQL database to use for Guacamole
                        authentication.
-END
+END`
```
I believe that trailing backtick here will result in the `END` failing to match the `<<END`, as the shell will be searching for a line that contains nothing but `END`. Can you confirm whether this works as expected?

WojoInc replied:
My very rough example compose file for a guacamole stack:

```yaml
version: "3.1"
services:
  daemon:
    image: guacamole/guacd:latest
    networks:
      - guac

  client:
    image: wojoinc/guac-test:latest
    networks:
      - guac
      - db
    ports:
      - "8081:8080"

    environment:
      GUACD_HOSTNAME: "daemon"
      GUACD_PORT: "4822"
      MYSQL_HOSTNAME: "db"
      #MYSQL_DATABASE_FILE: "/run/secrets/guac-mysql-db"
      MYSQL_PORT: "3306"
      MYSQL_USER_FILE: "/run/secrets/guac-mysql-user"
      MYSQL_PASSWORD_FILE: "/run/secrets/guac-mysql-password"

    secrets:
      - guac-mysql-db
      - guac-mysql-user
      - guac-mysql-password

  db:
    image: mariadb:latest
    networks:
      - db
    environment:
      MYSQL_ROOT_PASSWORD_FILE: "/run/secrets/guac-mysql-root-password"
      MYSQL_DATABASE: "guacamole"
      MYSQL_USER_FILE: "/run/secrets/guac-mysql-user"
      MYSQL_PASSWORD: "/run/secrets/guac-mysql-password"
      MYSQL_ROOT_HOST: "localhost"
    volumes:
      - "db_data:/var/lib/mysql"
      - "./initdb.sql:/docker-entrypoint-initdb.d/dump.sql"
    secrets:
      - guac-mysql-root-password
      - guac-mysql-db
      - guac-mysql-user
      - guac-mysql-password

networks:
  guac:

  db:

secrets:
  guac-mysql-root-password:
    external: true
  guac-mysql-db:
    external: true
  guac-mysql-user:
    external: true
  guac-mysql-password:
    external: true

volumes:
  db_data:
```

When omitting MYSQL_DATABASE or MYSQL_DATABASE_FILE, this causes the container to fail to spin up, and the logs present the error message from before calling associate_mysql() located at the bottom of the file,
```
FATAL: No authentication configured
-------------------------------------------------------------------------------
The Guacamole Docker container needs at least one authentication mechanism in
order to function, such as a MySQL database, PostgreSQL database, or LDAP
directory.  Please specify at least the MYSQL_DATABASE or POSTGRES_DATABASE
environment variables, or check Guacamole's Docker documentation regarding
configuring LDAP and/or custom extensions.
```

If you omit MYSQL_USER for example however, it does appear to read beyond the trailing backtick. I made the change to remove it, and I am rebuilding a test image now to see if it still works with that back tick removed. Thanks for letting me know!

necouchman commented on guacamole-docker/README.md:
```diff
@@ -45,6 +56,9 @@ the image will stop:
 2. `POSTGRES_USER` - The user that Guacamole will use to connect to PostgreSQL.
 3. `POSTGRES_PASSWORD` - The password that Guacamole will provide when connecting to PostgreSQL as `POSTGRES_USER`.
+4. `POSTGRES_DATABASE_FILE` - The path of the docker secret containing the name of database to use for Guacamole authentication.
+5. `POSTGRES_USER` - The path of the docker secret containing the name of the user that Guacamole will use to connect to PostgreSQL.
```
This should probably be `POSTGRES_USER_FILE` if it's the "path of the docker secret," no?

necouchman commented on guacamole-docker/README.md:
```diff
@@ -45,6 +56,9 @@ the image will stop:
 3. `POSTGRES_PASSWORD` - The password that Guacamole will provide when connecting to PostgreSQL as `POSTGRES_USER`.
+4. `POSTGRES_DATABASE_FILE` - The path of the docker secret containing the name of database to use for Guacamole authentication.
+5. `POSTGRES_USER` - The path of the docker secret containing the name of the user that Guacamole will use to connect to PostgreSQL.
+6. `POSTGRES_PASSWORD` - The path of the docker secret containing the password that Guacamole will provide when connecting to PostgreSQL as `POSTGRES_USER`.
```
This should probably be `POSTGRES_PASSWORD_FILE` if it's the "path of the docker secret," no?

necouchman commented on guacamole-docker/README.md:
```diff
@@ -90,6 +107,9 @@ the image will stop:
 2. `MYSQL_USER` - The user that Guacamole will use to connect to MySQL.
 3. `MYSQL_PASSWORD` - The password that Guacamole will provide when connecting to MySQL as `MYSQL_USER`.
+4. `MYSQL_DATABASE_FILE` - The path of the docker secret containing the name of database to use for Guacamole authentication.
+5. `MYSQL_USER` - The path of the docker secret containing the name of the user that Guacamole will use to connect to MySQL.
```
`MYSQL_USER` -> `MYSQL_USER_FILE`

necouchman commented on guacamole-docker/README.md:
```diff
@@ -90,6 +107,9 @@ the image will stop:
 3. `MYSQL_PASSWORD` - The password that Guacamole will provide when connecting to MySQL as `MYSQL_USER`.
+4. `MYSQL_DATABASE_FILE` - The path of the docker secret containing the name of database to use for Guacamole authentication.
+5. `MYSQL_USER` - The path of the docker secret containing the name of the user that Guacamole will use to connect to MySQL.
+6. `MYSQL_PASSWORD` - The path of the docker secret containing the password that Guacamole will provide when connecting to MySQL as `MYSQL_USER`.
```
`MYSQL_PASSWORD` -> `MYSQL_PASSWORD_FILE`

WojoInc commented:
My apologies, got busy with work and school. I'll go ahead and get on this
tomorrow and try to get everything cleaned up.

On Thu, Jan 17, 2019, 7:35 PM Virtually Nick <notifications@github.com>
wrote:

> @WojoInc <https://github.com/WojoInc> Just pinging on this to see if you
> can clean this up so we can merge this in?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/guacamole-client/pull/320#issuecomment-455395514>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEN6E-NfplxtRp3go86-8X96OcYxJ1J4ks5vESTvgaJpZM4WzVxy>
> .
>

WojoInc commented:
@necouchman Updated PR title and commit messages. Fixed the README issues as well and cleaned up the formatting. Please let me know if anything else looks off!
